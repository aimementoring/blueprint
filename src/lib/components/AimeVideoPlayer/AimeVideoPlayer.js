import React, { useState } from "react";
import PropTypes from "prop-types";
import ReactPlayer from "react-player";
import {
  componentPropTypes,
  defaultComponentPropTypes,
} from "../../utils/componentPropTypes";
import Modal from "../modal";

import styles from "./AimeVideoPlayer.module.scss";

// HLS info not sure if we want to add this might help performance? - https://www.dacast.com/blog/hls-streaming-protocol/

const AimeVideoPlayer = props => {
  const {
    url,
    mute,
    loop,
    title,
    theme,
    byLine,
    listType,
    imageUrl,
    controls,
    autoPlay,
    withModal,
    playListID,
    videoTitle,
    playsInPicture,
    backgroundVimeo,
    modalWithImage,
    videoDescription,
  } = props;
  const [showModal, setShowModal] = useState(false);
  const [urlWithModal, setUrlWithModal] = useState(url);

  const handleModal = () => {
    setShowModal(!showModal);
    setUrlWithModal(url);
  };

  const handleContextMenu = event => {
    event.preventDefault();
  };

  const customPlayIcon = (
    <>
      {withModal ? (
        <img
          className={styles.playButton}
          src="https://aime-website.s3.amazonaws.com/assets/images/play-btn-white.svg"
          alt="play video"
          onClick={handleModal}
        />
      ) : (
        <img
          src="https://aime-website.s3.amazonaws.com/assets/images/play-btn-white.svg"
          alt="play video"
        />
      )}
    </>
  );

  const videoPlayersConfig = {
    vimeo: {
      playerOptions: {
        loop,
        mute,
        title,
        controls,
        byline: byLine,
        frameborder: false,
        autoplay: autoPlay,
        background: backgroundVimeo,
      },
      preload: true,
    },
    youtube: {
      playerVars: {
        loop,
        rel: 0,
        controls,
        preload: true,
        color: "white",
        autoplay: autoPlay,
        frameborder: false,
        modestbranding: true,
        listType,
        list:
          playListID && playListID.indexOf("PL") > -1
            ? playListID
            : `PL${playListID}`,
      },
    },
    file: {
      /* attributes: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#Attributes */
      // tracks: [
      //   { kind: 'subtitles', src: 'subs/subtitles.en.vtt', srcLang: 'en', default: true },
      //   { kind: 'subtitles', src: 'subs/subtitles.ja.vtt', srcLang: 'ja' },
      //   { kind: 'subtitles', src: 'subs/subtitles.de.vtt', srcLang: 'de' },
      // ],
    },
  };

  const lightMode =
    withModal && !showModal ? withModal : !withModal && !showModal;

  const withPlaceHolderimage =
    imageUrl === "" ? lightMode : !showModal && imageUrl;

  const backGroundStyle = {
    background: `url(${imageUrl})`,
    backgroundSize: "cover",
    backgroundPosition: "center",
    backgroundRepeat: "no-repeat",
    backgroundAttachment: "fixed",
  };

  return (
    <div className={styles[`theme-${theme}`]}>
      {!showModal && (
        <div className={styles.playerContainer}>
          <div className={styles.playerBoarder}>
            <div
              className={styles.bgImage}
              style={imageUrl !== "" ? backGroundStyle : null}
            />
            {withModal && customPlayIcon}
            {videoTitle && (
              <span className={styles.videoTitle}>{videoTitle}</span>
            )}
            <ReactPlayer
              playsinline
              volume={0.7}
              width="100%"
              height="100%"
              url={url}
              light={withPlaceHolderimage}
              playing={!withModal}
              pip={playsInPicture}
              playIcon={lightMode && customPlayIcon}
              config={videoPlayersConfig}
              onContextMenu={handleContextMenu}
            />
            {videoDescription && (
              <div className={styles.videoDescription}>{videoDescription}</div>
            )}
          </div>
        </div>
      )}
      <Modal
        showModal={showModal}
        handleModal={handleModal}
        backGroundStyle={
          modalWithImage && imageUrl !== "" ? backGroundStyle : null
        }
        hideBodyOverflowY
      >
        <div className={styles.playerContainer}>
          <div className={styles.playerBoarder}>
            <ReactPlayer
              playsinline
              volume={0.7}
              width="100%"
              height="100%"
              playing={showModal}
              url={urlWithModal}
              config={videoPlayersConfig}
              onContextMenu={handleContextMenu}
            />
          </div>
        </div>
      </Modal>
    </div>
  );
};

AimeVideoPlayer.propTypes = {
  loop: PropTypes.bool,
  mute: PropTypes.bool,
  title: PropTypes.bool,
  byLine: PropTypes.bool,
  controls: PropTypes.bool,
  autoPlay: PropTypes.bool,
  withModal: PropTypes.bool,
  listType: PropTypes.string,
  imageUrl: PropTypes.string,
  playListID: PropTypes.string,
  videoTitle: PropTypes.string,
  playsInPicture: PropTypes.bool,
  modalWithImage: PropTypes.bool,
  backgroundVimeo: PropTypes.bool,
  url: PropTypes.string.isRequired,
  videoDescription: PropTypes.string,
  ...componentPropTypes,
};

AimeVideoPlayer.defaultProps = {
  loop: false,
  mute: false,
  title: false,
  imageUrl: "",
  listType: "",
  byLine: false,
  playListID: "",
  videoTitle: "",
  controls: false,
  autoPlay: false,
  withModal: false,
  videoDescription: "",
  playsInPicture: false,
  modalWithImage: false,
  backgroundVimeo: false,
  ...defaultComponentPropTypes,
};

export default AimeVideoPlayer;
