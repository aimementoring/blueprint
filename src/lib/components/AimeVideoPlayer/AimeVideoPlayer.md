```js
<AimeVideoPlayer
  videoTitle="This is a title"
  videoDescription="This is a video description"
  url="https://www.youtube.com/embed/FUWwQv1kklY"
  imageUrl="https://aime-website.s3.amazonaws.com/assets/images/know-aime/white-paper.jpg"
/>
```

Required prop:

- `url` = string

Optional props:

- `imageUrl` = string
- `listType` = string
- `withModal` = bool
- `videoTitle` = string
- `playListID` = string
- `modalWithImage` = bool
- `videoDescription` = string

videoPlayersConfig props (optional) most likley these will not need to be changed, but the option is available.

- `loop` = bool
- `mute` = bool
- `byLine` = bool
- `controls` = bool
- `autoPlay` = bool
- `listType` = string
- `playListID` = string
- `playsInPicture` = bool
- `backgroundVimeo` = bool

To play video with modal add this prop => `withModal`  
To change the placeholder image update this prop => `imageUrl` or for a more simple player remove custom placeholder image by removing `imageUrl` prop.

VIMEO:  
Depending on video owners settings on vimeo, controls can be hidden.  
Owner settings can be updated in vimeo webApp and controls can be hidden with paid versions of vimeo.  
controls for `vimeo.com/169599296` are hidden, change to this video `vimeo.com/22439234` the controls can't be hidden.

YOUTUBE:  
If you want other videos to play after the "inital video" (intial video = video in url prop)  
then add this prop => `listType="playlist"` with the playlist ID you want to play through, e.g. `playListID="PLjfNcXcq0TORvhTeJV_dgZrU2mjCeTqg2"`.

AIME:

Test URL's:  
youTube url: `https://www.youtube.com/embed/FUWwQv1kklY`  
vimeo url: `https://vimeo.com/169599296` | | `https://vimeo.com/22439234`
aime url: `https://s3-ap-southeast-2.amazonaws.com/video-api-production/encoding-output/161-8ea8db58-451f-43fb-a351-5f18cd0dcad4-1000290001/web.mp4`
