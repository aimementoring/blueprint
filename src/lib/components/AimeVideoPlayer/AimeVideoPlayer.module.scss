@import "../../styles/variables/variables.module.scss";

@import "../../styles/themes/themes.module.scss";

.playerContainer {
  display: flex;
  text-align: center;
  align-self: center;
  align-content: center;

  .playerBoarder {
    width: 58em;
    height: 32em;
    position: relative;
    padding: 0;
    border: solid #666464 1px;

    iframe {
      animation: fadein 2s ease;
    }

    .playButton {
      top: 50%;
      left: 50%;
      z-index: 100;
      cursor: pointer;
      position: absolute;
      transform: translate(-50%, -50%);
    }
  }
}

@keyframes fadein {

  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

/* Firefox < 16 */

@-moz-keyframes fadein {

  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

/* Safari, Chrome and Opera > 12.1 */

@-webkit-keyframes fadein {

  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

/* Internet Explorer */

@-ms-keyframes fadein {

  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}

/* Opera < 12.1 */

@-o-keyframes fadein {

  from {
    opacity: 0;
  }

  to {
    opacity: 1;
  }
}
