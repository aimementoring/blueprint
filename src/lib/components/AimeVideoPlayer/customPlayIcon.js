/* eslint-disable jsx-a11y/alt-text */
import React from "react";
import PropTypes from "prop-types";

const CustomPlayIcon = ({ withModal, onClick, stylesPlayButton }) => {
  const iconProps = {
    alt: "play video",
    src:
      "https://aime-website.s3.amazonaws.com/assets/images/play-btn-white.svg",
  };
  if (withModal) {
    iconProps.className = stylesPlayButton;
    iconProps.onClick = onClick;
  }
  return <img {...iconProps} />;
};

CustomPlayIcon.propTypes = {
  withModal: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  stylesPlayButton: PropTypes.object.isRequired,
};

CustomPlayIcon.defaultProps = {
  withModal: false,
};

export default CustomPlayIcon;
