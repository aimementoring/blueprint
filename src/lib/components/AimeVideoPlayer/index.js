import AimeVideoPlayer from './aimeVideoPlayer';

export default AimeVideoPlayer;
