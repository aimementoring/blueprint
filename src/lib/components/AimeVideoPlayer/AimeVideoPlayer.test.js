import React from "react";
import renderer from "react-test-renderer";
import AimeVideoPlayer from "./aimeVideoPlayer.js";

describe("AimeVideoPlayer", () => {
  it("renders properly", () => {
    const tree = renderer
      .create(<AimeVideoPlayer url="https://vimeo.com/169599296" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
