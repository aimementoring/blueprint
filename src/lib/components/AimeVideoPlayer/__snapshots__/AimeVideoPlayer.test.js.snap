// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AimeVideoPlayer renders properly 1`] = `
<div
  className="theme-base"
>
  <div
    className="playerContainer"
  >
    <div
      className="playerBoarder"
    >
      <div
        className="bgImage"
        style={null}
      />
      
      <div
        className="reactPlayer"
        onContextMenu={[Function]}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      >
        <div
          className="react-player__preview"
          onClick={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundImage": undefined,
              "backgroundPosition": "center",
              "backgroundSize": "cover",
              "cursor": "pointer",
              "display": "flex",
              "height": "100%",
              "justifyContent": "center",
              "width": "100%",
            }
          }
        >
          <img
            alt="play video"
            src="https://aime-website.s3.amazonaws.com/assets/images/play-btn-white.svg"
          />
        </div>
      </div>
      
    </div>
  </div>
</div>
`;
